zero_downtime_migrations/backend/schema.py

```diff
@@ -334,21 +340,57 @@ def _create_unique_constraint_from_index_sql(self, table, index_name):
             "index_name": index_name,
         }
 
+    def _check_index_sql(self, index_name):
+        return SQL_CHECK_INDEX_STATUS % {
+            "index_name": index_name,
+        }
+
+    def _check_valid_index(self, sql):
+        """
+        Return index_name if it's invalid
+        """
+        index_match = re.match(r'.* "(?P<index_name>.+)" ON .+', sql)
+        if index_match:
+            index_name = index_match.group('index_name')
+            check_index_sql = self._check_index_sql(index_name)
+            cursor = self.get_query_result(check_index_sql)
+            if self.parse_cursor_result(cursor=cursor):
+                return index_name
+
+    def _create_unique_failed(self, exc):
+        return (DJANGO_VERISON >= StrictVersion('2.1')
+                and 'could not create unique index' in repr(exc)
+                )
+
     def execute(self, sql, params=()):
         exit_atomic = False
         # Account for non-string statement objects.
         sql = str(sql)
 
-        if re.search('CREATE.+INDEX', sql):
+        if re.search('(CREATE|DROP).+INDEX', sql):
             exit_atomic = True
             if 'CONCURRENTLY' not in sql:
-                sql = sql.replace('CREATE INDEX', 'CREATE INDEX CONCURRENTLY')
+                sql = sql.replace('INDEX', 'INDEX CONCURRENTLY')
         atomic = self.connection.in_atomic_block
         if exit_atomic and atomic:
             self.atomic.__exit__(None, None, None)
-
-        super(ZeroDownTimeMixin, self).execute(sql, params)
-
+        try:
+            super(ZeroDownTimeMixin, self).execute(sql, params)
+        except django.db.utils.IntegrityError as exc:
+            # create unique index should be treated differently
+            # because it raises error, instead of quiet exit
+            if not self._create_unique_failed(exc):
+                raise
+
+        if exit_atomic and not self.collect_sql:
+            invalid_index_name = self._check_valid_index(sql)
+            if invalid_index_name:
+                # index was build, but invalid, we need to delete it
+                self.execute(self.sql_delete_index % {'name': invalid_index_name})
+                raise InvalidIndexError(
+                    'Unsuccessful attempt to create an index, fix data if needed and restart.'
```

Comment:
How data can be fixed by user? Is that obvious? For me it's not.

Reply:
it is mentioned in postgresql wiki, add some info to readme